hpct-ldap-server/src/managers/openldap.py

```diff
@@ -0,0 +1,491 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+#
+# managers/ldap.py
+"""Manager for server operator."""
+import logging
+import os
+import shutil
+import subprocess
+import tempfile
+
+logger = logging.getLogger(__name__)
+
+from charms.operator_libs_linux.v0 import apt
+from charms.operator_libs_linux.v1 import systemd
+
+from utils.filedata import FileData
+
+MIN_GID = 999
+MIN_UID = 999
+
+
+class OpenldapServerManager:
+    """Manager to provide LDAP server charm all functionality needed."""
+
+    packages = ["slapd", "ldap-utils", "sssd-ldap", "gnutls-bin", "ssl-cert"]
+    systemd_services = ["slapd"]
+
+    def __init__(self):
+        pass
+
+    def _add_base(self, admin_passwd, dcs) -> None:
+        """Define base for groups and users.
+
+        Parameters
+        ----------
+        admin_passwd : str
+            LDAP password.
+        dcs : str
+            Domain components.
+        """
+        base = [
+            f"dn: ou=People,{dcs}",
+            "objectClass: organizationalUnit",
+            "ou: People",
+            "",
+            f"dn: ou=Groups,{dcs}",
+            "objectClass: organizationalUnit",
+            "ou: Groups",
+        ]
+        base = "\n".join(base)
+
+        with open("/etc/ldap/basedn.ldif", "w") as f:
+            f.write(f"{base}")
+
+        binddn = f"cn=admin,{dcs}"
+        cmd = ["ldapadd", "-x", "-D", binddn, "-w", admin_passwd, "-f", "/etc/ldap/basedn.ldif"]
+
+        rc = subprocess.call(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.STDOUT)
+        if rc != 0:
+            raise Exception("Unable to add basedn ldif.")
+
+    def _split_domain(self, dc) -> str:
+        """Split domain components.
+
+        Parameters
+        ----------
+        dc : str
+            Domain components.
+        """
+        return ",".join([f"dc={x}" for x in dc.split(".")])
+
+    def add_user(
+        self,
+        admin_passwd=None,
+        domain=None,
+        gecos=None,
+        gid=None,
+        homedir=None,
+        shell=None,
+        passwd=None,
+        uid=None,
+        user=None,
+    ) -> None:
+        """Add user.
+
+        Parameters
+        ----------
+        admin_passwd : str
+            LDAP password.
+        domain : str
+            Domain name.
+        gecos : str
+            Real name.
+        gid : int
+            Group id.
+        homedir : str
+            Home directory path.
+        shell : str
+            Login shell.
+        passwd : str
+            User password.
+        uid : int
+            User id.
+        user : str
+            Username.
+        """
+        dcs = self._split_domain(domain)
+        if uid < MIN_UID or gid < MIN_GID:
+            logger.error("uid and gid must be below {MIN_UID}.")
+            return
+        if (
```

Comment:
This could check and early exit on `None`.

Reply:
Updated as suggested